### ISSUE TITLE
StopIteration issue

### ISSUE THREAD

**Issue by i4s** [RallyTools/RallyRestToolkitForPython#77]
Hi everybody, sorry for interruption, but I am a new in pyral and I always get this kind of error during creation connection. Could you help me, please?

```
# python2.7 ./rally1.py 
Traceback (most recent call last):
  File "rally1.py", line 2, in <module>
    rally = Rally("rally1.rallydev.com", user="username@domain.com", password="password")
  File "/usr/lib/python2.7/site-packages/pyral/restapi.py", line 228, in __init__
    self.contextHelper.check(self.server)
  File "/usr/lib/python2.7/site-packages/pyral/context.py", line 222, in check
    self._getDefaults(response)
  File "/usr/lib/python2.7/site-packages/pyral/context.py", line 262, in _getDefaults
    user = response.next()
  File "/usr/lib/python2.7/site-packages/pyral/rallyresp.py", line 264, in next
    raise StopIteration
StopIteration

# cat rally1.py 
from pyral import Rally
rally = Rally("rally1.rallydev.com", user="username@domain.com", password="password")

>>> pyral.__version__
(1, 1, 1)
>>> requests.__version__
'2.9.1'
```

I have tried script from here https://github.com/RallyTools/RallyRestToolkitForPython/issues/38 and have got the following answer:

```
$ ./basic.py 
https://rally1.rallydev.com/slm/webservice/v2.0/Subscription?fetch=true&pagesize=200&start=1
200
Request round-trip time:    1.00 secs

RESPONSE HEADER
{'Content-Length': '578', 'X-XSS-Protection': '1; mode=block', 'CF-RAY': '28bd9083110b0893-FRA', 'RallyRequestID': 'qd-app-061obufvs42w87131r2i6wogw1e.qd-app-06609127', 'Content-Encoding': 'gzip', 'Set-Cookie': '__cfduid=d0ac345c14dcb55664cdd586b52239d5b1459361910; expires=Thu, 30-Mar-17 18:18:30 GMT; path=/; domain=.rallydev.com; HttpOnly, JSESSIONID=qd-app-061obufvs42w87131r2i6wogw1e.qd-app-06;Path=/;Secure;HttpOnly, ZSESSIONID=p0vi1sXISKCjIHrySc1hPunLTDxdYU4SlfU0iRGH8XI;Path=/;Domain=rally1.rallydev.com;Secure;HttpOnly, SUBBUCKETID=590;Path=/;Domain=rally1.rallydev.com;Secure;HttpOnly, SERVERID=eef0381130fdee6ad45f9214be2b5a10994893f1; path=/', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains', 'Vary': 'Accept-Encoding', 'Server': 'cloudflare-nginx', 'Connection': 'keep-alive', 'ETag': '"0ce1684f024cdde0d0046a377ed4e27f9"', 'Cache-Control': 'private,max-age=0,must-revalidate', 'Date': 'Wed, 30 Mar 2016 18:18:30 GMT', 'P3P': 'CP="NON DSP COR CURa PSAa PSDa OUR NOR BUS PUR COM NAV STA"', 'Content-Type': 'application/json;  charset=utf-8', 'Expires': 'Thu, 01 Jan 1970 00:00:00 GMT'}

RESPONSE BODY
{"Subscription": {"_rallyAPIMajor": "2", "_rallyAPIMinor": "0", "_ref": "https://rally1.rallydev.com/slm/webservice/v2.0/subscription/3631859928", "_refObjectUUID": "60ab78be-acac-4d0c-872d-562d336de916", "_objectVersion": "41", "_refObjectName": "Company Name", "CreationDate": "2011-06-14T16:46:40.186Z", "_CreatedAt": "Jun 14, 2011", "ObjectID": 3631859928, "ObjectUUID": "60ab78be-acac-4d0c-872d-562d336de916", "VersionId": "41", "ApiKeysEnabled": true, "EmailEnabled": true, "ExpirationDate": null, "JSONPEnabled": true, "MaximumCustomUserFields": -1, "MaximumProjects": -1, "Modules": "Integration Hub,Rally Portfolio Manager,HP Integration,IBM Integration,Rally Idea Manager,Rally Quality Manager,Rally Support Manager,Rally Time Tracker,SSO,IP Restriction,Extension Whitelist", "Name": "Company Name", "PasswordExpirationDays": 0, "PreviousPasswordCount": 0, "ProjectHierarchyEnabled": true, "SessionTimeoutSeconds": null, "StoryHierarchyEnabled": true, "StoryHierarchyType": "Unlimited", "SubscriptionID": 6590, "SubscriptionType": "Unlimited", "Workspaces": {"_rallyAPIMajor": "2", "_rallyAPIMinor": "0", "_ref": "https://rally1.rallydev.com/slm/webservice/v2.0/Subscription/3631859928/Workspaces", "_type": "Workspace", "Count": 1}, "Errors": [], "Warnings": []}}
```

Thank you.

**Comment by klehman-rally** [comment]
On 2016-03-29 there was a modification to the WSAPI schema endpoint where metadata for Subscription and Workspace was no longer included in the results.  The exception traceback you observed is a bit different than most observations, but given the date that you filed the issue, it seems worthwhile to see if the WSAPI schema endpoint response structure modifications triggered the traceback you encountered.

The WSAPI change was toggled off after a few hours and pyral users observed normal behavior after that point.  If you are now still seeing the issue, one additional piece of diagnostic information would be most useful to examine.  To do that, I'd suggest using a Chrome browser REST extension like Advanced REST Client, Postman or REST Console.  Issue GET requests for https://rally1.rallydev.com/slm/webservice/v2.0/Subscription?fetch=true and https://rally1.rallydev.com/slm/webservice/v2.0/User?fetch=true.  For the request for User, you should see a response body (in formatted JSON) that looks very much like the following structure:{
    "User": {
        "_ref": "<ref_value>",
        "_refObjectUUID": "<UUID_value>f",
        "_objectVersion": "<integer>",
        "_refObjectName": "<your_first_name>",
        "CreationDate": "20xx-yy-zzThh:mm:ss.lllZ",
        "_CreatedAt": "Xxx dd, yyyy",
        "ObjectID": 999999999,
        "ObjectUUID": "<UUID_value>",
        "VersionId": "<integer>",
        "Subscription": {
            "_ref": "<ref_value>",
            "_refObjectUUID": "<UUID_value>",
            "_refObjectName": "<Your Subscription Name>",
            "_type": "Subscription"
        },
        "UserName": "<Your_user_name>",        
         ... several User related attributes elided...
        "EmailAddress": "<you>@<your_domain>",
        "FirstName": "<Your_first_name>",
        "LastName": "<Your_last_name>",
        "Role": "<Your_role>",
        "SubscriptionID": <integer>,
        "SubscriptionPermission": "<Some_role>",
        "TeamMemberships": {
            "_ref": "<ref_value>",
            "_type": "Project",
            "Count": <integer>  },
        "UserPermissions": {
            "_ref": "<ref_value>",
            "_type": "UserPermission",
            "Count": <integer>   },
        "UserProfile": {
            "_ref": "<ref_value>",
            "_refObjectUUID": "<UUID_value>",
            "_type": "UserProfile"  },
        "WorkspacePermission": "<Some_role>",
        "Errors": [],
        "Warnings": []
    }
}

**Comment by i4s** [comment]
Thank you for response.
I have started working with pyral 2 hours before I have posted the issue.
Yes, the problem is still reproducing for me.

For proposed requests I used Advanced REST client v4.12.8-stable.0 for Google-Chrome v47.0.2526.106 and have got the following responses:

```
> https://rally1.rallydev.com/slm/webservice/v2.0/Subscription?fetch=true
{"Subscription": {
    "_rallyAPIMajor": "2",
    "_rallyAPIMinor": "0",
    "_ref": "https://rally1.rallydev.com/slm/webservice/v2.0/subscription/3631859928",
    "_refObjectUUID": "60ab78be-acac-4d0c-872d-562d336de916",
    "_objectVersion": "41",
    "_refObjectName": "Company Name",
    "CreationDate": "2011-06-14T16:46:40.186Z",
    "_CreatedAt": "Jun 14, 2011",
    "ObjectID": 3631859928,
    "ObjectUUID": "60ab78be-acac-4d0c-872d-562d336de916",
    "VersionId": "41",
    "ApiKeysEnabled": true,
    "EmailEnabled": true,
    "ExpirationDate": null,
    "JSONPEnabled": true,
    "MaximumCustomUserFields": -1,
    "MaximumProjects": -1,
    "Modules": "Integration Hub,Rally Portfolio Manager,HP Integration,IBM Integration,Rally Idea Manager,Rally Quality Manager,Rally Support Manager,Rally Time Tracker,SSO,IP Restriction,Extension Whitelist",
    "Name": "Company Name",
    "PasswordExpirationDays": 0,
    "PreviousPasswordCount": 0,
    "ProjectHierarchyEnabled": true,
    "SessionTimeoutSeconds": null,
    "StoryHierarchyEnabled": true,
    "StoryHierarchyType": "Unlimited",
    "SubscriptionID": 6590,
    "SubscriptionType": "Unlimited",
    "Workspaces": {
        "_rallyAPIMajor": "2",
        "_rallyAPIMinor": "0",
        "_ref": "https://rally1.rallydev.com/slm/webservice/v2.0/Subscription/3631859928/Workspaces",
        "_type": "Workspace",
        "Count": 1
    },
    "Errors": [],
    "Warnings": []
}}
```

```
> https://rally1.rallydev.com/slm/webservice/v2.0/User?fetch=true
{"User": {
    "_rallyAPIMajor": "2",
    "_rallyAPIMinor": "0",
    "_ref": "https://rally1.rallydev.com/slm/webservice/v2.0/user/17715621255",
    "_refObjectUUID": "0c85a18c-ab24-471b-a43d-aa9b9514893d",
    "_objectVersion": "663",
    "_refObjectName": "My Name",
    "CreationDate": "2014-03-19T17:51:25.054Z",
    "_CreatedAt": "Mar 19, 2014",
    "ObjectID": 17715621255,
    "ObjectUUID": "0c85a18c-ab24-471b-a43d-aa9b9514893d",
    "VersionId": "663",
    "Subscription": {
        "_rallyAPIMajor": "2",
        "_rallyAPIMinor": "0",
        "_ref": "https://rally1.rallydev.com/slm/webservice/v2.0/subscription/3631859928",
        "_refObjectUUID": "60ab78be-acac-4d0c-872d-562d336de916",
        "_refObjectName": "Company Name",
        "_type": "Subscription"
    },
    "CostCenter": "None",
    "Department": "None",
    "Disabled": false,
    "DisplayName": "My Name",
    "EmailAddress": "My_Email@domain.com",
    "FirstName": "My First Name",
    "LandingPage": "/dashboard",
    "LastLoginDate": "2016-03-31T07:16:51.246Z",
    "LastName": "My Second Name",
    "LastPasswordUpdateDate": "2014-11-04T14:10:10.450Z",
    "LastSystemTimeZoneName": "Europe/Minsk",
    "MiddleName": null,
    "NetworkID": null,
    "OfficeLocation": "Minsk - BY",
    "OnpremLdapUsername": null,
    "Phone": null,
    "Planner": false,
    "Role": "Developer",
    "ShortDisplayName": null,
    "SubscriptionAdmin": false,
    "SubscriptionID": 6590,
    "SubscriptionPermission": "Workspace User",
    "TeamMemberships": {
        "_rallyAPIMajor": "2",
        "_rallyAPIMinor": "0",
        "_ref": "https://rally1.rallydev.com/slm/webservice/v2.0/User/17715621255/TeamMemberships",
        "_type": "Project",
        "Count": 2
    },
    "UserName": "my_email@domain.com",
    "UserPermissions": {
        "_rallyAPIMajor": "2",
        "_rallyAPIMinor": "0",
        "_ref": "https://rally1.rallydev.com/slm/webservice/v2.0/User/17715621255/UserPermissions",
        "_type": "UserPermission",
        "Count": 11
    },
    "UserProfile": {
        "_rallyAPIMajor": "2",
        "_rallyAPIMinor": "0",
        "_ref": "https://rally1.rallydev.com/slm/webservice/v2.0/userprofile/17715621257",
        "_refObjectUUID": "6aa249d3-cbb8-4e85-950d-b99514570e68",
        "_type": "UserProfile"
    },
    "WorkspacePermission": "Workspace User",
    "Errors": [0],
    "Warnings": [0]
}}
```

Thank you!

**Comment by klehman-rally** [comment]
This issue was fixed in v1.2.0.
